RandLAPACK/gpu_functions/rl_cuda_kernels.cuh

```diff
@@ -0,0 +1,934 @@
+    lower_bound = (lower_bound + numThreads - 1) / numThreads;
+    dim3 dimGrid(std::min(upper_bound, lower_bound), 1, 1);
+    void* kernelArgs[] = {(void*)&m, (void*)&n, (void*)&k, (void*)&A, (void*)&lda, (void*)&idx, (void*)&idx_copy};
+    cudaLaunchCooperativeKernel((void*)col_swap_gpu_kernel_sequential<T>, dimGrid, dimBlock, kernelArgs, 0, stream);
+    ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch col_swap_gpu sequential. " << cudaGetErrorString(ierr))
+        abort();
+    }
+    cudaFreeAsync(idx_copy, stream);
+    ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to deallocate for col_swap_gpu. " << cudaGetErrorString(ierr))
+        abort();
+    }
+#endif
+}
+
+/// Positions columns of A in accordance with idx vector of length k.
+/// idx array modified ONLY within the scope of this function.
+template <typename T>
+void col_swap_gpu(
+    cudaStream_t stream, 
+    int64_t m, 
+    int64_t n, 
+    int64_t k,
+    T* A, 
+    int64_t lda,
+    T* A_cpy,
+    int64_t ldac,
+    int64_t const* idx
+) {
+#ifdef USE_CUDA
+
+    dim3 threadsPerBlock(32, 8);
+    dim3 blocksPerGrid((k + threadsPerBlock.x - 1) / threadsPerBlock.x,
+                         (m + threadsPerBlock.y - 1) / threadsPerBlock.y);
+    col_swap_gpu_kernel<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(m, n, k, A, lda, A_cpy, ldac, idx);
+#endif
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch col_swap_gpu with parallel pivots. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+/// Positions columns of A in accordance with idx vector of length k.
+/// idx array modified ONLY within the scope of this function.
+template <typename T>
+void col_swap_gpu(
+    cudaStream_t stream, 
+    int64_t m,  
+    int64_t k,
+    int64_t* J, 
+    int64_t const* idx
+) {
+#if 1
+    std::vector<int64_t> idx_copy(k);
+    auto j = std::make_unique<int64_t[]>(m);
+    cudaMemcpyAsync(idx_copy.data(), idx, sizeof(int64_t) * k, cudaMemcpyDeviceToHost, stream);
+    cudaMemcpyAsync(j.get(), J, sizeof(int64_t) * m, cudaMemcpyDeviceToHost, stream);
+    cudaStreamSynchronize(stream);
+
+    RandLAPACK::util::col_swap<T>(m, k, j.get(), std::move(idx_copy));
+
+    cudaMemcpyAsync(J, j.get(), sizeof(int64_t) * m, cudaMemcpyHostToDevice, stream);
+    // must add this to avoid dangling reference during async copy
+    cudaStreamSynchronize(stream);
+#else
+#ifdef USE_CUDA
+    int64_t* idx_copy;
+    cudaMallocAsync(&idx_copy, sizeof(int64_t) * k, stream);
+    vec_ell_swap_gpu<T><<<1, 512, 0, stream>>>(m, k, J, idx, idx_copy);
+#endif
+    cudaFreeAsync(idx_copy, stream);
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch col_swap_gpu. " << cudaGetErrorString(ierr))
+        abort();
+    }
```

Comment:
Please remove the comments and replace with a line that throws an error.

Reply:
Fixed.